atom/browser/api/atom_api_web_contents.h

```diff
@@ -82,18 +82,29 @@ class WebContents : public mate::TrackableObject<WebContents>,
   using PrintToPDFCallback =
       base::Callback<void(v8::Local<v8::Value>, v8::Local<v8::Value>)>;
 
-  // Create from an existing WebContents.
-  static mate::Handle<WebContents> CreateFrom(
-      v8::Isolate* isolate,
-      content::WebContents* web_contents);
-  static mate::Handle<WebContents> CreateFrom(
+  // Create a new WebContents and return the V8 wrapper of it.
+  static mate::Handle<WebContents> Create(v8::Isolate* isolate,
+                                          const mate::Dictionary& options);
+
+  // Create a new V8 wrapper for an existing |web_content|.
+  //
+  // The lifetime of |web_contents| will be managed by this class.
+  static mate::Handle<WebContents> CreateAndTake(
       v8::Isolate* isolate,
       content::WebContents* web_contents,
```

Comment:
Can this be `std::unique_ptr` to signify it, the content layer has made it clear with the lifetime of `content::WebContents`, right now we release and take ownership because of existing signature, would good if it was just a `std::move` call.

Reply:
This is limited by how we wrap WebContents.

The `CreateAndTake` is called in `WebContentsCreated` which passes a raw pointer. The `unique_ptr` is passed in `AddNewContents`, but it is called after `WebContentsCreated` and the code assumes the wrapper has already been created by then.

So basically between `WebContentsCreated` and `AddNewContents` the pointers is being managed by two `unique_ptr`s, and Chromium gives up the ownership after `AddNewContents`.

This problem exists before this PR, this refactor just makes it stand out. A decent way to solve this is to not emit JS event in `AddNewContents`, but that would be another refactoring work.